source/memcached-store.ts

```diff
@@ -0,0 +1,194 @@
+		this.expiration = options.windowMs / 1000 // [in seconds]
+	}
+
+	/**
+	 * Method to prefix the keys with the given text.
+	 *
+	 * @param key {string} - The key.
+	 *
+	 * @returns {string} - The text + the key.
+	 */
+	prefixKey(key: string): string {
+		return `${this.prefix}${key}`
+	}
+
+	/**
+	 * Method to increment a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client.
+	 *
+	 * @returns {IncrementResponse} - The number of hits and reset time for that client.
+	 */
+	async increment(key: string): Promise<IncrementResponse> {
+		const response = (await this.updateKey(
+			this.prefixKey(key),
+			+1,
+		)) as IncrementResponse // We can cast this for sure because we're passing delta as +1
+		return response
+	}
+
+	/**
+	 * Method to decrement a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client
+	 */
+	async decrement(key: string): Promise<void> {
+		await this.updateKey(this.prefixKey(key), -1)
+	}
+
+	/**
+	 * Method to reset a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client.
+	 */
+	async resetKey(key: string): Promise<void> {
+		const prefixedKey = this.prefixKey(key)
+
+		// Delete the record for that key.
+		await new Promise<void>((resolve, reject) => {
+			this.client.del(prefixedKey, (error) => {
+				if (error) {
+					reject(error)
+					return
+				}
+
+				resolve()
+			})
+		})
+	}
+
+	private async updateKey(
+		key: string,
+		delta: number,
+	): Promise<IncrementResponse | void> {
+		// Find the existing record for that key.
+		let record = await new Promise<ClientRecord>((resolve, reject) => {
+			this.client.get(key, (error, data) => {
+				if (error) {
+					reject(error)
+					return
+				}
+
+				resolve(data as ClientRecord)
+			})
+		})
+
+		if (record === undefined && delta < 0) {
+			// If the record does not exist, and we are supposed to decrement the key,
+			// we don't need to do anything, so we return.
+			return
+		}
+
+		if (record === undefined) {
+			// If a record does not exist, and must be incremented, then add a record
+			// for that key.
+			record = await new Promise<ClientRecord>((resolve, reject) => {
+				const payload: ClientRecord = { hits: 1, time: Date.now() }
+
+				this.client.add(key, payload, this.expiration, (error) => {
+					if (error) {
+						reject(error)
+						return
+					}
+
+					resolve(payload)
+				})
+			})
+		} else {
+			// If it does exist, simply change the hit count by `delta`.
+			record = await new Promise<ClientRecord>((resolve, reject) => {
+				const payload: ClientRecord = { ...record, hits: record.hits + delta }
+				const expiration = Math.floor(
+					// = window - (time elapsed since start of window) [in seconds]
+					this.expiration - (Date.now() - payload.time) / 1000,
+				)
+
+				this.client.replace(key, payload, expiration, (error) => {
+					if (error) {
+						reject(error)
+						return
+					}
+
+					resolve(payload)
+				})
+			})
+		}
```

Comment:
I'd love to use the new meta arithmetic command, but for right now, what about this strategy:

1. call incr on the key, if it succeeds:
    * do a `get` on the key, suffixed with '_expiry'
    * return both values
2. if the incr call failed because the key doesn't exist:
    * set the key to 1
    * set `${key}_expiry` to the expiration time
    * return both of those values

This is roughly what the previous version did, except we're adding the expiration time in a new key. It also has the benefit of being backwards compatible with existing users.

We can always add the meta arithmetic version later as an enhancement later on.

Reply:
Also, I think that should go into the `increment` method, `decrement`should just call decr on the key (ignoring both the expiration time and any failures), and we should drop the `updateKey` method.

Reply:
I was working on the meta arithmetic command, but the private `.command` function doesn't seem to return the output from the redis command.

Your idea sounds good, I'll try implementing it.